libc/stdio_file.c

```diff
@@ -0,0 +1,785 @@
+
+
+FILE *fd, *fd2;
+char stdpath[] = "stdio_file_test";
+char c, buf[20];
+int filedes;
+char toolongpath[PATH_MAX + 16];
+
+/* 
+Tets group for:
+fopen, fclose,
+fdopen, freopen
+*/
+TEST_GROUP(fopenfclose);
+
+TEST_SETUP(fopenfclose)
+{
+	memset(toolongpath, 'a', PATH_MAX + 15);
+	toolongpath[PATH_MAX + 15] = '\0';
+}
+
+
+TEST_TEAR_DOWN(fopenfclose)
+{ /* empty */
+}
+
+
+int checkdescriptor(FILE *fd)
+{
+	return (fd == NULL) ? 0 : !fclose(fd);
+}
+
+
+void assert_fopen_error(char *path, char *opts, int errnocode)
+{
+	fd = fopen(path, opts);
+	TEST_ASSERT_FALSE(checkdescriptor(fd));
+	TEST_ASSERT_EQUAL_INT(errnocode, errno);
+}
+
+
+void assert_fopen_success(char *path, char *opts)
+{
+	fd = fopen(path, opts);
+	TEST_ASSERT_TRUE(checkdescriptor(fd));
+}
+
+
+TEST(fopenfclose, fopenfclose_file)
+{
+	/* not existing file opening without creating */
+	assert_fopen_error(stdpath, "r", ENOENT);
+	assert_fopen_error(stdpath, "r+", ENOENT);
+	/* opening file with creation */
+	assert_fopen_success(stdpath, "w");
+	assert_fopen_success(stdpath, "a");
+	assert_fopen_success(stdpath, "w+");
+	assert_fopen_success(stdpath, "a+");
+	/* opening existing file for read */
+	assert_fopen_success(stdpath, "r");
+	assert_fopen_success(stdpath, "r+");
+}
+
+
+TEST(fopenfclose, fopenfclose_opendir)
+{
+	/* open directory */
+	assert_fopen_success("/dev/", "r");
+	assert_fopen_error("/dev/", "w", EISDIR);
+}
+
+TEST(fopenfclose, fopenfclose_zeropath)
+{
+	/* open null or zero path */
+	assert_fopen_error("", "r", ENOENT);
+	assert_fopen_error(NULL, "r", EINVAL);
+	assert_fopen_error("", "w", ENOENT);
+	assert_fopen_error(NULL, "w", EINVAL);
+}
+
+TEST(fopenfclose, fopenfclose_wrongflags)
+{
+	/* open with no flags/wrong flags/null flags */
+	assert_fopen_error(stdpath, "", EINVAL);
+	assert_fopen_error(stdpath, "phoenix-rtos", EINVAL);
+	assert_fopen_error(stdpath, NULL, EINVAL);
+}
+
+TEST(fopenfclose, fopenfclose_toolongname)
+{
+	/* open file with too long name */
+	assert_fopen_error(toolongpath, "w", ENAMETOOLONG);
+}
+
+
+TEST(fopenfclose, freopen_file)
+{
+	fd = fopen(stdpath, "w");
+	/* freopen() on opened file */
+	fd2 = freopen(stdpath, "w", fd);
+	{
+		TEST_ASSERT_NOT_NULL(fd);
+		TEST_ASSERT_NOT_NULL(fd2);
+		TEST_ASSERT_TRUE(fd == fd2);
+	}
+	fclose(fd);
+}
+
+
+TEST(fopenfclose, fdopen_file)
+{
+	fd = fopen(stdpath, "r");
+	{
+		TEST_ASSERT_NOT_NULL(fd);
+		filedes = fileno(fd);
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(0, filedes);
+		fd2 = fdopen(filedes, "r");
+		TEST_ASSERT_NOT_NULL(fd2);
+	}
+	fclose(fd);
+}
+
+/* 
+Tets group for:
+- fwrite, fread
+- putc, fputc,
+- getc, fgetc,
+- putchar, getchar, < need threads
+- ungetc,
+- puts, fputs, fgets
+*/
+TEST_GROUP(getput);
+
+TEST_SETUP(getput)
+{ /* empty */
+}
+
+
+TEST_TEAR_DOWN(getput)
+{ /* empty */
+}
+
+
+TEST(getput, fwritefread_basic)
+{
+	/* write some data to file using fwrite(), read it using fread(), assert end of file */
+	memset(buf, 0, 20);
+	fd = fopen(stdpath, "w+");
+	{
```

Comment:
Why extra block?

Reply:
As to visually represent file operations. Each block represents operations on an open file, thus it is clear to see if it is opened and closed properly. Just a visual gimmick not to leave open files behind.

Reply:
Yeah, I've noticed the pattern later, forgot to delete this comment.